core/src/main/java/hudson/PluginFirstClassLoader2.java

```diff
@@ -0,0 +1,167 @@
+package hudson;
+
+import edu.umd.cs.findbugs.annotations.CheckForNull;
+import hudson.util.CompoundEnumeration;
+import java.io.IOException;
+import java.net.URL;
+import java.util.Enumeration;
+import java.util.Objects;
+import jenkins.util.URLClassLoader2;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+
+/**
+ * Class loader that consults the system class loader, the plugin's {@code WEB-INF/lib/*.jar} and
+ * {@code WEB-INF/classes} directories, and the Jenkins core class loader (in that order).
+ *
+ * <p>To use this class loader, set {@code pluginFirstClassLoader} to {@code true} in the {@code
+ * maven-hpi-plugin} configuration.
+ *
+ * @author Basil Crow
+ * @since TODO
+ */
+@Restricted(NoExternalUse.class)
+public class PluginFirstClassLoader2 extends URLClassLoader2 {
```

Comment:
would it be possible to add a test for this (there is already a [test ](https://github.com/jenkinsci/jenkins/blob/e8f2ae4e4c1ae63ae48d18f2a7b7134f317c7f7f/test/src/test/java/hudson/PluginManagerTest.java#L551-L565)for the older plugin first)

Reply:
> there is already a [test](https://github.com/jenkinsci/jenkins/blob/e8f2ae4e4c1ae63ae48d18f2a7b7134f317c7f7f/test/src/test/java/hudson/PluginManagerTest.java#L551-L565) for the older plugin first classloader

I verified that the existing `PluginManagerTest#findResourceForPluginFirstClassLoader` passes with both `hudson.ClassicPluginStrategy.useAntClassLoader=true` (`PluginFirstClassLoader`) and `hudson.ClassicPluginStrategy.useAntClassLoader=false` (`PluginFirstClassLoader2`). Due to how early `ClassicPluginStrategy` is loaded during the tests, it was too difficult to get a test to set the flag early enough to cover both cases (even with `FlagRule`), but if and when we ever change the default value of the flag, the existing test will exercise the corresponding class loader.

Reply:
> but if and when we ever change the default value of the flag, the existing test will exercise the corresponding class loader.

👍 

FWIW `RealJenkinsRule` may help.